### ISSUE TITLE
What is the purpose of Load(ORM/ODM)MetadataSubscriber?

### ISSUE THREAD

**Issue by bendavies** [Sylius/Sylius#5040]
Hi!

I'm using Sylius Resource Bundle to serve a rest API.

Simply, `LoadORMMetadataSubscriber` is breaking some the associations in my custom entities, specifically the `unsetAssociationMappings` method.

Before trying to fix the behaviour, I need to understand what it is for!
What is the purpose of these listeners?

Thanks

**Comment by pjedrzejewski** [comment]
Example: in Sylius we provide PaymentMethod resource, unless you specify your own class for this resource, our class ``Sylius\Component\Payment\Model\PaymentMethod`` is an mapped-superclass. We convert it to an entity if you have not specified your own implementation. Also, some of the relations cannot be defined on mapped-superclasses, like Cart.items. To save the developer the hassle of redefining these in their custom implementations - we move them automatically via this listener as well.

To sum up:

* It turns non-customized resources into entities.
* Moves associations when necessary.

Let me know what does it break exactly, we have fixed a number of issues with that listener, hopefully this is something easy too. ;)

**Comment by bendavies** [comment]
I have Entities setup like so:

`Resource extends SomeMappedSuperClass`
`SomeMappedSupperClass extends BaseClass`

`BaseClass` has a `OneToOne` relation, which `LoadORMMetadataSubscriber` unsets.

Since these Subscribers are fixing Sylius specific issues, they should act on a whitelist of Entities, and not all entities?

**Comment by pjedrzejewski** [comment]
They currently act on all resources. Agree we should limit it to base Sylius entities! 👍

**Comment by liverbool** [comment]
Hi, After #5047 merged i got errors for entity mapping.

<img width="1146" alt="screen shot 2559-05-19 at 00 13 27" src="https://cloud.githubusercontent.com/assets/206804/15368194/bd7ba44e-1d56-11e6-9c70-682f8f028512.png">

Example config.

```yaml
Vcare\Bundle\CoreBundle\Model\ProductVariantImage:
    type: entity
    table: sylius_product_variant_image

    fields:
        position:
            type: integer
            nullable: false
```

How to fix?

**Comment by liverbool** [comment]
For brief check on `\Sylius\Bundle\ResourceBundle\EventListener\ORMMappedSuperClassSubscriber`  i think we should check `isSyliusClass()` only for `unsetAssociationMappings()` ?

**Comment by bendavies** [comment]
@liverbool I'm not sure. If you look at the [old version](https://github.com/Sylius/Sylius/blob/e2c5d8ebe688061ed366a97747057f8da1f9b44e/src/Sylius/Bundle/ResourceBundle/EventListener/LoadORMMetadataSubscriber.php), it ran for all classes. Now, it runs for only Sylius classes.

Can you roll back and try figure out what is wrong with the new version?

Maybe the new version should run on entities if they are Sylius entities, or if one of their parents is a Sylius entity? @pjedrzejewski ?

Since I broke this, if you can share a fork demonstrating the problem, i'm happy to have a look too.

**Comment by liverbool** [comment]
I'm fine when not use `isSyliusClass()`, Not sure #4745 can related issue for this.

**Comment by bendavies** [comment]
yes but that is needed so the class does not break NON sylius entities.
So there is still some work to do...
If you can push a fork, i can help

**Comment by liverbool** [comment]
Hmm, my fork is private, As you see my example config.

There are my steps.

**Model**
```
Vcare\Bundle\CoreBundle\Model\ProductVariantImage extends Sylius\Component\Core\Model\ProductVariantImage
```

**Mapping**
```yaml
Vcare\Bundle\CoreBundle\Model\ProductVariantImage:
    type: entity
    table: sylius_product_variant_image

    fields:
        position:
            type: integer
            nullable: false
```

**Resource Override**
```yaml
sylius_core:
    resources:
        product_variant_image:
            classes:
                model: Vcare\Bundle\CoreBundle\Model\ProductVariantImage
                interface: Vcare\Bundle\CoreBundle\Model\ProductVariantImageInterface
```

**#4745 Fix**
```yaml
doctrine:
    orm:
        resolve_target_entities:
            Sylius\Component\Core\Model\ProductVariantImageInterface: Vcare\Bundle\CoreBundle\Model\ProductVariantImage

```

**Comment by pjedrzejewski** [comment]
@bendavies I think you are right about running it on entities which are Sylius or have Sylius parent.

**Comment by bendavies** [comment]
@liverbool a minimum reproduction case fork would be appreciated.

**Comment by okwinza** [comment]
@bendavies just try to checkout fresh master then override any of Sylius entity with our own and try to run doctrine validation after. Shouldn't take too long.

**Comment by bendavies** [comment]
@liverbool i followed your steps and could not replicated. please provide a fork.

**Comment by pjedrzejewski** [comment]
@liverbool Usually, you should not reconfigure the ``interface`` changing the model is enough. I just tried your config and it works: https://github.com/pjedrzejewski/Sylius/tree/reproduce-issue-with-mapping

**Comment by liverbool** [comment]
Hi Guys, i ran into @pjedrzejewski fork and saw everything is ok, But my repo still bugy i have no idea why i still got errors, i make my fork to be public and push simple bugy code to https://github.com/intbizth/vcare/tree/remove-dos-resource

@bendavies @pjedrzejewski Can you help to check my repo please 😢 😢

**Comment by bendavies** [comment]
@liverbool are you aware you are ignoring composer.lock? This is bad practise for non libraries. you could just fix to the old version of resource bundle if you weren't ignoring it and your problem would go away until we fixed it.

Anyway, i'm taking a look now.

**Comment by bendavies** [comment]
@liverbool please make your project private again. I have it now.

**Comment by pjedrzejewski** [comment]
@bendavies Let me know if you are working on the fix, that's the last thing before tag, if you do not have time, I can have a look too, thanks! :)

**Comment by liverbool** [comment]
@bendavies thanks you and you are right about `composer.lock`, so it too
many conflict in my team but i will bring it back to git track again to
solve bigger issue /:

**Comment by bendavies** [comment]
@pjedrzejewski I'm am. PR soon.
